### PR TITLE
[INLONG-4890][Sort] Support TubeMQ connector

### LINKED ISSUE #4890: [Feature][Sort] Support TubeMQ connector
### Description

TubeMQ Connector supports Flink Table API and SQL.
This feature currently supports only reads.

TubeMQ introduce：https://inlong.apache.org/docs/next/modules/tubemq/overview

### Use case

_No response_

### Are you willing to submit PR?

- [X] Yes, I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by GanfengTan**
feature: Add TubeMQ connector
1.It can be used through the Flink Table API and SQL.
2.This feature currently supports only reads.

TubeMQ introduce：https://inlong.apache.org/docs/next/modules/tubemq/overview

- Fixes #4890 

### Motivation

1.Extend support for MQ types.
2.Use the large-scale data capabilities of TubeMQ.

### Modifications

1.Add TubeMQ consumers .
2.Extend Flink's dynamic table.

### Verifying this change

- [x] This change is a trivial rework/code cleanup without any test coverage.

### Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a follow-up issue for adding the documentation

### CODE REVIEW

File: inlong-sort/sort-connectors/tubemq/src/main/java/org/apache/inlong/sort/tubemq/FlinkTubeMQConsumer.java [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+    public void run(SourceContext<T> ctx) throws Exception {
+
+        Instant lastConsumeInstant = Instant.now();
+
+        while (running) {
+
+            ConsumerResult consumeResult = messagePullConsumer.getMessage();
+            if (!consumeResult.isSuccess()) {
+                if (!(consumeResult.getErrCode() == 400
+                        || consumeResult.getErrCode() == 404
```
Comment (EMsnap) [inline inlong-sort/sort-connectors/tubemq/src/main/java/org/apache/inlong/sort/tubemq/FlinkTubeMQConsumer.java:257]: magic num, please extract constant

File: inlong-sort/sort-connectors/tubemq/src/main/java/org/apache/inlong/sort/tubemq/FlinkTubeMQConsumer.java [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+    public FlinkTubeMQConsumer(
+            String masterAddress,
+            String topic,
+            TreeSet<String> tidSet,
+            String consumerGroup,
+            DeserializationSchema<T> deserializationSchema,
+            Configuration configuration,
+            String sessionKey
+    ) {
+        checkNotNull(masterAddress,
```
Comment (healchow) [inline inlong-sort/sort-connectors/tubemq/src/main/java/org/apache/inlong/sort/tubemq/FlinkTubeMQConsumer.java:165]: Those lines can be merged into one line.

File: inlong-sort/sort-connectors/tubemq/src/main/java/org/apache/inlong/sort/tubemq/table/TubeMQDynamicTableFactory.java [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+        final ReadableConfig tableOptions = helper.getOptions();
+
+        final DecodingFormat<DeserializationSchema<RowData>> valueDecodingFormat =
+                getValueDecodingFormat(helper);
+
+        // validate all options
+        helper.validate();
+
+        helper.validateExcept(PROPERTIES_PREFIX);
```
Comment (thesumery) [inline inlong-sort/sort-connectors/tubemq/src/main/java/org/apache/inlong/sort/tubemq/table/TubeMQDynamicTableFactory.java:121]: here nested format is support ? ForExample format: inlongMsg , innert format is json.

Comment (GanfengTan) [reply inlong-sort/sort-connectors/tubemq/src/main/java/org/apache/inlong/sort/tubemq/table/TubeMQDynamicTableFactory.java:121]: Internal customization is not supported and needs to be extended by adding parameters. Currently flink formats are supported json, CSV

**Comment by thesumery** [comment]
It's better to use Flink new Source API. More detail to read this reference: https://cwiki.apache.org/confluence/display/FLINK/FLIP-27%3A+Refactor+Source+Interface

File: inlong-sort/sort-connectors/tubemq/src/main/java/org/apache/inlong/sort/tubemq/FlinkTubeMQConsumer.java [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+                    T record =
+                            deserializationSchema.deserialize(message.getData());
+                    records.add(record);
+                }
+            }
+
+            synchronized (ctx.getCheckpointLock()) {
+
+                for (T record : records) {
+                    ctx.collect(record);
```
Comment (EMsnap) [inline inlong-sort/sort-connectors/tubemq/src/main/java/org/apache/inlong/sort/tubemq/FlinkTubeMQConsumer.java:293]: can checkpoint lock only lock currentoffsets ? this lock is used in several places which may cause degration in term s of performance

File: inlong-sort/sort-connectors/tubemq/src/main/java/org/apache/inlong/sort/tubemq/FlinkTubeMQConsumer.java [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+                }
+
+                continue;
+            }
+
+            List<Message> messageList = consumeResult.getMessageList();
+            lastConsumeInstant = Instant.now();
+
+            List<T> records = new ArrayList<>();
+            if (messageList != null) {
```
Comment (Oneal65) [inline inlong-sort/sort-connectors/tubemq/src/main/java/org/apache/inlong/sort/tubemq/FlinkTubeMQConsumer.java:280]: whether the case  of  messageList.length is 0  is  correct?

File: inlong-sort/sort-connectors/tubemq/pom.xml [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        <dependency>
+            <groupId>org.apache.inlong</groupId>
+            <artifactId>tubemq-client</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.inlong</groupId>
+            <artifactId>sort-common</artifactId>
+            <version>${project.version}</version>
+        </dependency>
```
Comment (gong) [inline inlong-sort/sort-connectors/tubemq/pom.xml:43]: where use sort-common?

File: inlong-sort/sort-connectors/tubemq/pom.xml [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+                    <execution>
+                        <id>shade-flink</id>
+                        <phase>package</phase>
+                        <goals>
+                            <goal>shade</goal>
+                        </goals>
+                        <configuration>
+                            <artifactSet>
+                                <includes>
+                                    <include>org.apache.inlong:tubemq-client</include>
```
Comment (gong) [inline inlong-sort/sort-connectors/tubemq/pom.xml:65]: dependencies of `tubemq-client` will miss

File: inlong-sort/sort-connectors/tubemq/pom.xml (hunk 6 above)
Comment (GanfengTan) [reply inlong-sort/sort-connectors/tubemq/pom.xml:43]: Error introductions have been shaved off.

File: inlong-sort/sort-connectors/tubemq/src/main/java/org/apache/inlong/sort/tubemq/table/TubeMQDynamicTableFactory.java [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+    public DynamicTableSource createDynamicTableSource(Context context) {
+        final TableFactoryHelper helper = FactoryUtil.createTableFactoryHelper(this, context);
+
+        final ReadableConfig tableOptions = helper.getOptions();
+
+        final DecodingFormat<DeserializationSchema<RowData>> valueDecodingFormat =
+                getValueDecodingFormat(helper);
+
+        // validate all options
+        helper.validate();
```
[the remark this answers is not in the document]
Comment (GanfengTan) [reply inlong-sort/sort-connectors/tubemq/src/main/java/org/apache/inlong/sort/tubemq/table/TubeMQDynamicTableFactory.java:119]: Has been removed

File: inlong-sort/sort-connectors/tubemq/src/main/java/org/apache/inlong/sort/tubemq/FlinkTubeMQConsumer.java (hunk 5 above)
Comment (GanfengTan) [reply inlong-sort/sort-connectors/tubemq/src/main/java/org/apache/inlong/sort/tubemq/FlinkTubeMQConsumer.java:280]: No this case

File: inlong-sort/sort-connectors/tubemq/pom.xml (hunk 7 above)
Comment (GanfengTan) [reply inlong-sort/sort-connectors/tubemq/pom.xml]: done

File: inlong-sort/sort-connectors/tubemq/src/main/java/org/apache/inlong/sort/tubemq/FlinkTubeMQConsumer.java (hunk 4 above)
Comment (GanfengTan) [reply inlong-sort/sort-connectors/tubemq/src/main/java/org/apache/inlong/sort/tubemq/FlinkTubeMQConsumer.java:264]: https://github.com/apache/inlong/pull/4891#issuecomment-1177005578

Yes, you are right.
This is going to be an improvement term.Base on @thesumery advice.

**Comment by GanfengTan** [comment]
@thesumery Thanks your advice, as an improvement.

**Comment by healchow** [comment]
@GanfengTan Hello, is there any other code that needs to be submitted?

**Comment by GanfengTan** [comment]
@healchow  There is no code to submit.

**Review by EMsnap — APPROVED** [review]

**Review by yunqingmoswu — APPROVED** [review]

**Review by gong — APPROVED** [review]

File: inlong-sort/sort-connectors/tubemq/src/main/java/org/apache/inlong/sort/tubemq/table/TubeMQDynamicTableFactory.java [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+    @Override
+    public DynamicTableSource createDynamicTableSource(Context context) {
+        final TableFactoryHelper helper = FactoryUtil.createTableFactoryHelper(this, context);
+
+        final ReadableConfig tableOptions = helper.getOptions();
+
+        final DecodingFormat<DeserializationSchema<RowData>> valueDecodingFormat = getValueDecodingFormat(helper);
+
+        // validate all options
+        helper.validate();
```
Comment (thesumery) [inline inlong-sort/sort-connectors/tubemq/src/main/java/org/apache/inlong/sort/tubemq/table/TubeMQDynamicTableFactory.java:113]: Here table options with 'properties.*' will be validate and here will throw exception?
Any other source with their own properties validate their own options will use `validateExcept` to except some prefix.

Comment (GanfengTan) [reply inlong-sort/sort-connectors/tubemq/src/main/java/org/apache/inlong/sort/tubemq/table/TubeMQDynamicTableFactory.java:113]: This is an extension and currently has no impact.

**Review by gosonzhang — APPROVED** [review]

**Pull request merged**